### PR TITLE
sensors: create vehicle_accelerometer module

### PR DESCRIPTION
**Pull request by dagar**
Builds on https://github.com/PX4/Firmware/pull/12596.

This adds a high rate, low latency, vehicle body frame, filtered, bias corrected acceleration for use in various controllers throughout the code base. The structure is analogous to vehicle_angular_velocity.

### CODE REVIEW

File: msg/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +122,7 @@ set(msg_files
 	uavcan_parameter_value.msg
 	ulog_stream.msg
 	ulog_stream_ack.msg
+	vehicle_accelerometer.msg
```
[the remark this answers is not in the document]
Comment (jlecoeur) [reply msg/CMakeLists.txt]: Option A:
module `VehicleAcceleration` -> topic `vehicle_acceleration`
module `VehicleAngularVelocity` -> topic `vehicle_angular_velocity`

Option B:
module `VehicleAccelerometer` -> topic `vehicle_acceleration`
module `VehicleGyroscope` -> topic `vehicle_angular_velocity`

File: src/modules/sensors/vehicle_accelerometer/VehicleAccelerometer.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+{
+	perf_begin(_cycle_perf);
+	perf_count(_interval_perf);
+
+	// update corrections first to set _selected_accel
+	SensorCorrectionsUpdate();
+
+	sensor_accel_s sensor_accel;
+
+	if (_sensor_accel_sub[_selected_accel].update(&sensor_accel)) {
```
Comment (jlecoeur) [inline src/modules/sensors/vehicle_accelerometer/VehicleAccelerometer.cpp]: `_selected_accel` *may* still be -1 if the sensor_accel callback registration failed

File: msg/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (dagar) [reply msg/CMakeLists.txt]: I agree, but I didn't want to cause confusion with vehicle acceleration not including gravity.

File: src/modules/land_detector/MulticopterLandDetector.cpp [hunk 3]
```diff
@@ -126,17 +127,11 @@ bool MulticopterLandDetector::_get_freefall_state()
...
-		return false;
-	}
-
-	float acc_norm = _sensor_bias.accel_x * _sensor_bias.accel_x
-			 + _sensor_bias.accel_y * _sensor_bias.accel_y
-			 + _sensor_bias.accel_z * _sensor_bias.accel_z;
-	acc_norm = sqrtf(acc_norm);	//norm of specific force. Should be close to 9.8 m/s^2 when landed.
+	// norm of specific force. Should be close to 9.8 m/s^2 when landed.
+	const matrix::Vector3f accel{_vehicle_accelerometer.xyz};
+	const float acc_norm = sqrtf(accel(0) * accel(0) + accel(1) * accel(1) + accel(2) * accel(2));
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/land_detector/MulticopterLandDetector.cpp]: Much better.

File: src/modules/land_detector/MulticopterLandDetector.cpp [hunk 4]
```diff
@@ -126,17 +127,11 @@ bool MulticopterLandDetector::_get_freefall_state()
 		return false;
 	}
 
-	if (_sensor_bias.timestamp == 0) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/land_detector/MulticopterLandDetector.cpp]: The new topic should be available before the land detector is started, but it's safer to keep.

File: src/modules/sensors/vehicle_accelerometer/VehicleAccelerometer.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/sensors/vehicle_accelerometer/VehicleAccelerometer.cpp]: I'll make it an unsigned integer (initialized to zero), and force the first update.

File: msg/vehicle_accelerometer.msg [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
+uint64 timestamp		# time since system start (microseconds)
+
+uint64 timestamp_sample		# the timestamp of the raw data (microseconds)
+
+float32[3] xyz			# Bias corrected acceleration in body axis (in m/s^2)
```
[the remark this answers is not in the document]
Comment (dagar) [reply msg/vehicle_accelerometer.msg]: Done.

**Comment by dagar** [comment]
I've applied most of your suggestions, including renaming to `vehicle_acceleration` with a mention of gravity in the comment. The goal of the original naming (vehicle_accelerometer) was to try and prevent potential confusion. Thoughts?

**Comment by jlecoeur** [comment]
I agree that it can get confusing. There are three different things people may want: measured acceleration (body frame, including gravity), body linear acceleration (body frame, excluding gravity) and linear acceleration (ground frame, excluding gravity, aka derivative of NED velocity). Finding unambiguous names is difficult here. At least the comment in the message definition is clear.

**Pull request merged**